Add modifications to support compiling on iOS.

freakboy3742 opened the pull request:
I've been working on improving iOS support in the Python ecosystem. I'm currently working on porting NumPy as an example of a high profile, non-trivial Python package. 

This PR makes a number of small changes to support building Numpy binary wheels for iOS:

* Adds some missing details and utility methods for the `ios`, `tvos` and `watchos` system types
* Modifies the Linker interface to allow them to have an awareness of the system they are linking for. This is required because the Apple linker requires different semantics depending on whether you're targeting macOS or iOS (and, presumably, tvOS and watchOS), but the underlying linker is actually the same. Some other APIs on the linker interface are provided an `env` argument; adding `system` as a property of the linker seemed a lot less invasive.
* Modifies the Apple Linker to disable the use of `-undefined,dynamic_lookup`. The use of `-undefined,dynamic_lookup` is listed as deprecated in the iOS compiler, and using this flag will raise a deprecation warning.
* Modifies the Apple Linker to use `-dynamiclib` when linking iOS modules, rather than `-bundle`. iOS apps have very strict guidelines on what can be included; `bundle` content is rejected by the App Store pre-validation process. The only non-static content allowed in an iOS app is a dynamically linked library, distributed as a Framework that is bundled with the app. This change has already been merged as part of mesonbuild/meson#14340.
* Modifies the Accelerate identification process used by BLAS/LAPACK to check for an appropriate iOS version. The LP64 features in Accelerate are marked as being available from iOS 16.4.

With the exception of the `blas_lapack` identification mechanism, these changes have been submitted upstream to Meson as mesonbuild/meson#14444.

freakboy3742 committed:
Add numpy specific extensions for iOS builds.

mattip commented:
> With the exception of the blas_lapack identification mechanism, these changes have been submitted upstream to Meson as https://github.com/mesonbuild/meson/pull/14444.

There are some CI failures, are you sure this is passing upstream? Do you need to rebase off numpy `main` or something?

freakboy3742 commented:
Yes - it's passing CI upstream... the CI failures here are very odd - there's a bunch of linting errors in files I haven't touched; test failures that appear unrelated to changes that I've made; and packaging errors caused by workflows that have been deprecated.

It's the end of my workday here; I can take a closer look in the morning. Does the main branch of this repo currently pass CI? It looks like it hasn't been run on
main for six months, and even then   I'm seeing a lot of failures in the Actions history.

rgommers commented:
No worries about the CI failures. Meson's CI config isn't set up to work well on repos other than github.com/mesonbuild/meson. We can ignore them. The way I've been going about testing PRs here is to integrate them in NumPy itself and ensure NumPy's CI is happy.

rgommers reviewed (commented):
Thanks @freakboy3742!

The BLAS-related changes here LGTM. I'll keep an eye on https://github.com/mesonbuild/meson/pull/14444, which is actively being reviewed right now. I think once that lands we can integrate this and update the bundled version in NumPy. The second bullet point sounds like the largest change, it'd be good to not diverge from upstream Meson there.

bonzini committed:
environment: handle all iOS variants as xnu

All of iOS, tvOS, visionOS, watchOS use the XNU kernel.  Report that and also
make them return true for is_darwin() which is really more like "is_xnu()".

Co-authored-by: Russell Keith-Magee <russell@keith-magee.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
linkers: pass system to DynamicLinker.__init__ for Darwin linkers

Apple linkers need to use different arguments on macOS and iOS-like platforms.
Pass the system to the constructor so that it can be examined.

Co-authored-by: Russell Keith-Magee <russell@keith-magee.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
linkers: apple: fix shared module args

iOS should not use -undefined,dynamic_lookup; whereas macOS should revert from
-dynamiclib to -bundle, as was the case before commit cfb5a48e0 ("linkers: darwin:
do not use -bundle for shared_modules", 2025-03-24), because Postgres wants to
use -bundle_loader and that is only possible together with -bundle.

Co-authored-by: Russell Keith-Magee <russell@keith-magee.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

freakboy3742 commented:
An update - it looks like mesonbuild/meson#14444 has been closed in favor of mesonbuild/meson#14481 by a Meson collaborator; that PR has all the same changes, but has a couple of other modifications that look like they could be useful for Rust compilation.

rgommers commented:
Thanks for the update - subscribed to the new PR. 

It looks like that is moving and will land well before the next NumPy release (2.3.0, branch date is at least 6 weeks off), so there is time - but if it hasn't landed in a month and you need this for the next NumPy release, then I'm willing to merge this temporarily into our forked meson @freakboy3742, so do feel free to ask if needed.

eli-schwartz committed:
linkers: fix regression when using lld after iOS changes

```
$ LDFLAGS="-fuse-ld=lld" meson setup builddir/
[...]
    linker = lld_cls(
        compiler, for_machine, comp_class.LINKER_PREFIX, override, system=system, version=v)
TypeError: LLVMDynamicLinker.__init__() got an unexpected keyword argument 'system'
```

Fixes regression in commit cece1a7e9992a3bbcc35f90777ba22ba9d62b538. We
pass system=system to the linker construction, which worked fine for
Apple LLVM LD64 as that inherited `__init__` from DynamicLinker. But
non-Apple LLD had an overridden `__init__` which wasn't adapted.

Even if it is thrown away and never used, since LLVM isn't an Apple
linker, we still need to obey the argument contract.

freakboy3742 commented:
Looks like that new PR has been merged; but there's been an almost immediate follow up (mesonbuild/meson#14486) to correct a bug in some of the extra changes. 

@rgommers Procedurally - how would you like to me to handle this? Do you want to manage the merge with upstream yourself, or would you like to update this PR to reflect the minimal set of changes needed for compatibility?

rgommers commented:
Great!

> Procedurally - how would you like to me to handle this? Do you want to manage the merge with upstream yourself, or would you like to update this PR to reflect the minimal set of changes needed for compatibility?

If you could update this PR to the minimal set of changes needed, that'd be helpful. Upgrading to upstream `master` will take more time than I have at the moment, I'm expecting conflicts in both the BLAS and SIMD code for that - that's for some other time.

freakboy3742 commented:
@rgommers I've just force-updated this branch to contain the 4 cherry picked commits from upstream (48fc3ec8fa25c0cb001603f025ab516e8f0df1e2, cece1a7e9992a3bbcc35f90777ba22ba9d62b538, 7a43b6e318d7a35f0ce250e80a9a04b6b2983f05 and 944456b5f9b4428853f231368e2baea979042c46) that implement the iOS changes, plus an additional commit that adds the blas/accelerate piece for iOS support.

I've tested [my numpy branch](https://github.com/numpy/numpy/pull/28759) with this set of changes, plus mesonbuild/meson-python#731; with those changes, numpy builds. The wheels work in simple manual testing, but the test suite isn't running yet - iOS can't use subprocesses, which the numpy test suite seems to hinge on. I'm still working on that.

rgommers reviewed (approved):
I now tested this in combination with gh-18. NumPy CI jobs seem happy, and the changes LGTM. So in it goes. Thanks again @freakboy3742!

I'll follow up with an update in the main NumPy repo, and will also incorporate the BLAS changes into https://github.com/mesonbuild/meson/pull/10921 so they don't get lost later on.

freakboy3742 commented:
@rgommers Thanks! As a heads up, some of the discussion on [the corresponding meson-python PR](https://github.com/mesonbuild/meson-python/pull/731) suggests there might be an additional change coming; some of the changes that I had in meson-python may be more appropriate up stream as part of Meson's Python dependency checks. I'm working on those updates at the moment.

rgommers commented:
Thanks for the heads up. Sure, we can include those if they're still needed after the CPython bug fix you mention there.

freakboy3742 commented:
I've submitted the extra changes upstream; see mesonbuild/meson#14541. I'll generate an updated PR for this repo if/when that one is merged.

freakboy3742 commented:
@rgommers Upstream meson patch has been merged; see #21 for the backport to this repo.
